devenv/dev-dashboards/datasource-testdata/new_features_in_v8.json

```diff
@@ -107,8 +102,7 @@
               "mode": "absolute",
               "steps": [
                 {
-                  "color": "green",
-                  "value": null
```

Comment:
think this value needs to be null for this to work (replaced with -Infinity at runtime )

Reply:
Seems like a perfect concrete case to discuss the general logic i outlined in the OP! So, this is a case where explicitly having the `null` maps to `-Infinity`. What does an absent `"value"` key map to, then?

Reply:
Oh wait, is it this way because `-Infinity` is a magical Javascript value that has no representation in JSON? (This doesn't actually change anything, i think, but it is an interesting twist)

Reply:
> Is it this way because -Infinity is a magical Javascript value that has no representation in JSON

yes

Reply:
value does not really allow nulls, but -Infinity is serialized as null :( 
https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/thresholds.ts#L1 

Looks like we do not have any check on null actually but just force the first step to have a value of -Infinitty https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/state/getPanelOptionsWithDefaults.ts#L192 

So removing this value should be fine

Reply:
Ah! ok, thank you, the flow that produces this particular `null` makes more sense now.

For clarity, [here's](https://github.com/grafana/grafana/pull/37594/files#diff-920013f8bb5c45e5ac9590a88656b7f093813f1d63d60121baeaea806480d22cR97-R98) the WIP schema declaration corresponding to this field; i just updated the comment to reflect this new knowledge. (N.B. that file is not its eventual intended canonical location).

This will be an interesting case wrt how much information we want to express in schema. It wouldn't be wrong to leave it pretty much as-is in the PR, and certainly that's a fine place to start. But it would be possible to encode the "first threshold is different" semantics, too. Something along these lines:

```cue
#FirstThreshold: {
  color: string
  state?: string  
}
#Threshold: {
  color: string
  state?: string
  value: number
}
#ThresholdsConfig: {
  steps: [#FirstThreshold, ...#Threshold]
  mode: #ThresholdsMode
}
```

It's not obvious that we make things much better by doing this. And i have not yet plumbed the depths of how CUE handles relatively complex structures like this in how it checks backwards compatibility (essential to providing scuemata's promised invariants). That, and the fact that the `steps` list is intended to have the invariant that the list should be sorted ascending by `#Threshold.value` - which we can write a constraint to check in CUE, but that won't actually perform the sort - suggests that the above might better live in a set of helpers we'll eventually produce (a successor to grafonnet).